bigframes/operations/__init__.py

```diff
@@ -126,86 +139,101 @@ def _convert_expr_input(
 
 
 # Operation Factories
-def create_unary_op(name: str) -> UnaryOp:
+def create_unary_op(
+    name: str, type_rule: op_typing.OpTypeRule = op_typing.INPUT_TYPE
+) -> UnaryOp:
+    def output_type(self, *input_types: dtypes.ExpressionType) -> dtypes.ExpressionType:
+        return type_rule.output_type(*input_types)
+
     return dataclasses.make_dataclass(
         name,
-        [("name", typing.ClassVar[str], name)],  # type: ignore
+        [("name", typing.ClassVar[str], name), ("output_type", typing.ClassVar[typing.Callable], output_type)],  # type: ignore
         bases=(UnaryOp,),
         frozen=True,
     )()
 
 
-def create_binary_op(name: str) -> BinaryOp:
+def create_binary_op(
+    name: str, type_rule: op_typing.OpTypeRule = op_typing.Supertype()
+) -> BinaryOp:
+    def output_type(self, *input_types: dtypes.ExpressionType) -> dtypes.ExpressionType:
```

Comment:
This seems repetitive all create_*_op methods. Would it make sense to define outside once and reuse? Or even define in a base `class Op` and derive the *naryOp classes from that?

Reply:
deduplicated into OpTypeRule class itself. Maybe should jsut have OpTypeRule be a Callable anyways...